File: Databases/databases.jl
# ============================================================================
# Databases - Julia Implementation
# ============================================================================

# Agenda
# Thanks to Prof. Cosma Shalizi (CMU Statistics) for this material
# - What databases are, and why
# - SQL
# - Interfacing Julia and SQL

# ============================================================================
# What are Databases?
# ============================================================================

# A record is a collection of fields
# A table is a collection of records which all have the same fields (with different values)
# A database is a collection of tables

# ============================================================================
# Connecting Julia to SQL Databases
# ============================================================================

# There are several packages for connecting Julia to SQL databases:
# - SQLite.jl: for SQLite databases (file-based, no server needed)
# - MySQL.jl: for MySQL/MariaDB databases
# - LibPQ.jl: for PostgreSQL databases
# - DBInterface.jl: general interface for various databases
# - DataFrames.jl: for working with tabular data

using SQLite
using DataFrames
using DBInterface

# ============================================================================
# Example 1: SQLite Database Connection
# ============================================================================

# Create a connection to SQLite database
# If file doesn't exist, it will be created
db = SQLite.DB("my_database.db")

# Create a sample table
DBInterface.execute(db, """
    CREATE TABLE IF NOT EXISTS students (
        id INTEGER PRIMARY KEY,
        name TEXT NOT NULL,
        age INTEGER,
        grade REAL
    )
""")

# Insert data into the table
DBInterface.execute(db, """
    INSERT INTO students (name, age, grade) VALUES 
    ('Alice', 20, 85.5),
    ('Bob', 22, 92.0),
    ('Charlie', 21, 78.5),
    ('Diana', 23, 88.0)
""")

# Query data from the table
result = DBInterface.execute(db, "SELECT * FROM students") |> DataFrame
println("All students:")
println(result)

# Query with filtering
high_performers = DBInterface.execute(db, 
    "SELECT name, grade FROM students WHERE grade > 85") |> DataFrame
println("\nHigh performers:")
println(high_performers)

# Query with aggregation
avg_grade = DBInterface.execute(db, 
    "SELECT AVG(grade) as average_grade FROM students") |> DataFrame
println("\nAverage grade:")
println(avg_grade)

# ============================================================================
# Example 2: Using SQLite.jl Directly
# ============================================================================

# Query returns a DataFrame directly
students_df = SQLite.Query(db, "SELECT * FROM students") |> DataFrame
println("\nStudents DataFrame:")
println(students_df)

# Query with filtering
high_performers_df = SQLite.Query(db, 
    "SELECT name, grade FROM students WHERE grade > 85") |> DataFrame
println("\nHigh performers DataFrame:")
println(high_performers_df)

# ============================================================================
# Example 3: Working with Existing DataFrames
# ============================================================================

# Create a sample DataFrame
students_data = DataFrame(
    id = [1, 2, 3, 4, 5],
    name = ["Alice", "Bob", "Charlie", "Diana", "Eve"],
    age = [20, 22, 21, 23, 20],
    grade = [85.5, 92.0, 78.5, 88.0, 95.5]
)

# Write DataFrame to database
SQLite.load!(students_data, db, "students", replace=true)

# Read entire table back
students_from_db = SQLite.Query(db, "SELECT * FROM students") |> DataFrame
println("\nStudents from database:")
println(students_from_db)

# Get list of tables in the database
tables = SQLite.tables(db) |> DataFrame
println("\nTables in database:")
println(tables)

# Get column information for a table
columns = SQLite.columns(db, "students") |> DataFrame
println("\nColumns in students table:")
println(columns)

# ============================================================================
# Example 4: MySQL Database Connection
# ============================================================================

# For MySQL databases (requires MySQL server running)
# using MySQL
# 
# # Connect to MySQL database
# conn = MySQL.connect(
#     host="localhost",
#     port=3306,
#     user="username",
#     passwd="password",
#     db="my_database"
# )
# 
# # Run queries
# result = MySQL.query(conn, "SELECT * FROM my_table") |> DataFrame
# 
# # Close connection
# MySQL.disconnect(conn)

# ============================================================================
# Example 5: PostgreSQL Database Connection
# ============================================================================

# For PostgreSQL databases (requires PostgreSQL server running)
# using LibPQ
# 
# # Connect to PostgreSQL database
# conn = LibPQ.Connection("""
#     host=localhost
#     port=5432
#     dbname=my_database
#     user=username
#     password=password
# """)
# 
# # Run queries
# result = LibPQ.execute(conn, "SELECT * FROM my_table") |> DataFrame
# 
# # Close connection
# close(conn)

# ============================================================================
# Common SQL Operations in Julia
# ============================================================================

# 1. SELECT - Retrieve data
result = DBInterface.execute(db, "SELECT * FROM students") |> DataFrame
println("\n1. SELECT all:")
println(result)

result = DBInterface.execute(db, 
    "SELECT name, grade FROM students WHERE age > 20") |> DataFrame
println("\n1. SELECT with WHERE:")
println(result)

# 2. INSERT - Add new records
DBInterface.execute(db, 
    "INSERT INTO students (name, age, grade) VALUES ('Frank', 24, 89.0)")
println("\n2. INSERT: Added Frank")

# 3. UPDATE - Modify existing records
DBInterface.execute(db, 
    "UPDATE students SET grade = 90.0 WHERE name = 'Frank'")
println("\n3. UPDATE: Updated Frank's grade")

# 4. DELETE - Remove records
DBInterface.execute(db, 
    "DELETE FROM students WHERE name = 'Frank'")
println("\n4. DELETE: Removed Frank")

# 5. ORDER BY - Sort results
result = DBInterface.execute(db, 
    "SELECT * FROM students ORDER BY grade DESC") |> DataFrame
println("\n5. ORDER BY grade DESC:")
println(result)

# 6. GROUP BY - Aggregate data
result = DBInterface.execute(db, 
    "SELECT age, AVG(grade) as avg_grade FROM students GROUP BY age") |> DataFrame
println("\n6. GROUP BY age:")
println(result)

# 7. JOIN - Combine tables (example with two tables)
# First create a courses table
DBInterface.execute(db, """
    CREATE TABLE IF NOT EXISTS courses (
        student_id INTEGER,
        course_name TEXT,
        credits INTEGER
    )
""")

DBInterface.execute(db, """
    INSERT INTO courses (student_id, course_name, credits) VALUES 
    (1, 'Math', 3),
    (1, 'Physics', 4),
    (2, 'Chemistry', 3),
    (3, 'Biology', 4)
""")

# Join students and courses
result = DBInterface.execute(db, """
    SELECT s.name, c.course_name, c.credits 
    FROM students s 
    JOIN courses c ON s.id = c.student_id
""") |> DataFrame
println("\n7. JOIN students and courses:")
println(result)

# 8. COUNT, SUM, AVG, MIN, MAX
result = DBInterface.execute(db, """
    SELECT 
        COUNT(*) as total_students,
        AVG(grade) as avg_grade,
        MIN(grade) as min_grade,
        MAX(grade) as max_grade
    FROM students
""") |> DataFrame
println("\n8. Aggregate functions:")
println(result)

# ============================================================================
# Using Parameterized Queries (Prevent SQL Injection)
# ============================================================================

# SQLite.jl supports parameterized queries
name_to_find = "Alice"
stmt = SQLite.Stmt(db, "SELECT * FROM students WHERE name = ?")
DBInterface.execute(stmt, [name_to_find])
result = DataFrame(stmt)
println("\nParameterized query result:")
println(result)

# Alternative using bind parameters
result = DBInterface.execute(db, 
    "SELECT * FROM students WHERE name = ?", [name_to_find]) |> DataFrame
println("\nParameterized query (alternative):")
println(result)

# ============================================================================
# Using Prepared Statements for Better Performance
# ============================================================================

# Prepare statement once, execute multiple times
stmt = SQLite.Stmt(db, "INSERT INTO students (name, age, grade) VALUES (?, ?, ?)")

# Execute with different parameters
DBInterface.execute(stmt, ["Grace", 22, 87.0])
DBInterface.execute(stmt, ["Henry", 21, 91.5])

println("\nInserted multiple rows using prepared statement")

# Verify the insertions
result = DBInterface.execute(db, 
    "SELECT * FROM students WHERE name IN ('Grace', 'Henry')") |> DataFrame
println(result)

# ============================================================================
# Transactions
# ============================================================================

# Begin a transaction
SQLite.transaction(db) do
    DBInterface.execute(db, 
        "INSERT INTO students (name, age, grade) VALUES ('Isabel', 20, 88.5)")
    DBInterface.execute(db, 
        "INSERT INTO students (name, age, grade) VALUES ('Jack', 23, 93.0)")
    # Automatically commits if no exception, rolls back if exception occurs
end

println("\nTransaction completed")

# ============================================================================
# Working with DataFrames and SQL
# ============================================================================

using Statistics

# Query to DataFrame
students_df = SQLite.Query(db, "SELECT * FROM students") |> DataFrame

# Use DataFrames operations (alternative to SQL)
# Filter
high_performers = filter(row -> row.grade > 85, students_df)
println("\nHigh performers (using DataFrames):")
println(high_performers)

# Sort
sorted_students = sort(students_df, :grade, rev=true)
println("\nSorted by grade (using DataFrames):")
println(sorted_students)

# Group by and aggregate
using DataFramesMeta
grouped = @chain students_df begin
    groupby(:age)
    @combine(:avg_grade = mean(:grade))
end
println("\nGrouped by age (using DataFrames):")
println(grouped)

# ============================================================================
# Query Builder Pattern
# ============================================================================

# Building queries programmatically
function build_query(table::String; 
                     columns::Vector{String}=["*"], 
                     where::Union{String,Nothing}=nothing,
                     order_by::Union{String,Nothing}=nothing)
    query = "SELECT $(join(columns, ", ")) FROM $table"
    
    if !isnothing(where)
        query *= " WHERE $where"
    end
    
    if !isnothing(order_by)
        query *= " ORDER BY $order_by"
    end
    
    return query
end

# Use the query builder
query = build_query("students", 
                   columns=["name", "grade"], 
                   where="grade > 85",
                   order_by="grade DESC")
println("\nBuilt query: $query")
result = DBInterface.execute(db, query) |> DataFrame
println(result)

# ============================================================================
# Handling NULL Values
# ============================================================================

# Insert a record with NULL
DBInterface.execute(db, 
    "INSERT INTO students (name, age, grade) VALUES ('Karen', NULL, 85.0)")

# Query with NULL handling
result = DBInterface.execute(db, 
    "SELECT name, age, grade FROM students WHERE age IS NULL") |> DataFrame
println("\nStudents with NULL age:")
println(result)

# Clean up
DBInterface.execute(db, "DELETE FROM students WHERE name = 'Karen'")

# ============================================================================
# Batch Operations
# ============================================================================

# Insert multiple rows efficiently
new_students = DataFrame(
    name = ["Liam", "Mia", "Noah"],
    age = [22, 21, 23],
    grade = [88.5, 91.0, 87.5]
)

SQLite.load!(new_students, db, "students")
println("\nBatch insert completed")

# ============================================================================
# Database Inspection
# ============================================================================

# Get database schema
schema = SQLite.Query(db, "SELECT sql FROM sqlite_master WHERE type='table'") |> DataFrame
println("\nDatabase schema:")
println(schema)

# Get table info
table_info = SQLite.Query(db, "PRAGMA table_info(students)") |> DataFrame
println("\nTable info for students:")
println(table_info)

# Get indices
indices = SQLite.Query(db, "PRAGMA index_list(students)") |> DataFrame
println("\nIndices for students:")
println(indices)

# ============================================================================
# Best Practices
# ============================================================================

# 1. Use parameterized queries to prevent SQL injection
#    DBInterface.execute(db, "SELECT * FROM students WHERE name = ?", [name])

# 2. Use transactions for multiple related operations
#    SQLite.transaction(db) do
#        # multiple operations
#    end

# 3. Use prepared statements for repeated queries
#    stmt = SQLite.Stmt(db, "SELECT * FROM students WHERE age = ?")
#    DBInterface.execute(stmt, [age_value])

# 4. Use DataFrames for easy data manipulation
#    df = SQLite.Query(db, query) |> DataFrame

# 5. Handle NULL values explicitly
#    Check for missing values in DataFrames

# 6. Close database connections when done
#    SQLite.DB connections are automatically managed in Julia

# 7. Use batch operations for better performance
#    SQLite.load! for bulk inserts

# ============================================================================
# Performance Tips
# ============================================================================

# 1. Create indices for frequently queried columns
DBInterface.execute(db, "CREATE INDEX IF NOT EXISTS idx_grade ON students(grade)")

# 2. Use ANALYZE to update statistics
DBInterface.execute(db, "ANALYZE")

# 3. Use EXPLAIN QUERY PLAN to understand query performance
explain = SQLite.Query(db, "EXPLAIN QUERY PLAN SELECT * FROM students WHERE grade > 85") |> DataFrame
println("\nQuery plan:")
println(explain)

# 4. Use bulk operations instead of row-by-row
#    SQLite.load! is much faster than individual inserts

# ============================================================================
# Cleanup
# ============================================================================

# SQLite connections in Julia are automatically cleaned up
# But you can explicitly close if needed
# SQLite.close(db)

# Optionally, remove the database file
# rm("my_database.db")

println("\n" * "="^80)
println("Summary")
println("="^80)
println("""
A database is basically a way of dealing efficiently with lots of 
potentially huge DataFrames

SQL is the standard language for telling databases what to do, 
especially what queries to run

Everything in an SQL query is something we've practiced already in Julia:
- Filtering (WHERE clause = filter())
- Aggregation (GROUP BY = groupby())
- Merging (JOIN = join())
- Ordering (ORDER BY = sort())

Workflow:
1. Connect Julia to the database
2. Send it an SQL query
3. Analyze the returned DataFrame

Key packages:
- SQLite.jl: SQLite databases
- MySQL.jl: MySQL databases
- LibPQ.jl: PostgreSQL databases
- DBInterface.jl: Common interface
- DataFrames.jl: Tabular data manipulation

More information at: http://www.stat.cmu.edu/~cshalizi/statcomp/14/
""")

File: Databases/databases.py
# ============================================================================
# Databases - Python Implementation
# ============================================================================

# Agenda
# Thanks to Prof. Cosma Shalizi (CMU Statistics) for this material
# - What databases are, and why
# - SQL
# - Interfacing Python and SQL

# ============================================================================
# What are Databases?
# ============================================================================

# A record is a collection of fields
# A table is a collection of records which all have the same fields (with different values)
# A database is a collection of tables

# ============================================================================
# Connecting Python to SQL Databases
# ============================================================================

# There are several packages for connecting Python to SQL databases:
# - sqlite3: for SQLite databases (built-in, no installation needed)
# - pymysql/mysql-connector-python: for MySQL/MariaDB databases
# - psycopg2: for PostgreSQL databases
# - sqlalchemy: general ORM and interface for various databases
# - pandas: has built-in SQL support via read_sql and to_sql

import sqlite3
import pandas as pd
from sqlalchemy import create_engine, text
import os

# ============================================================================
# Example 1: SQLite Database Connection (using sqlite3)
# ============================================================================

# Create a connection to SQLite database
# If file doesn't exist, it will be created
conn = sqlite3.connect('my_database.db')
cursor = conn.cursor()

# Create a sample table
cursor.execute("""
    CREATE TABLE IF NOT EXISTS students (
        id INTEGER PRIMARY KEY,
        name TEXT NOT NULL,
        age INTEGER,
        grade REAL
    )
""")

# Insert data into the table
cursor.execute("""
    INSERT INTO students (name, age, grade) VALUES 
    ('Alice', 20, 85.5),
    ('Bob', 22, 92.0),
    ('Charlie', 21, 78.5),
    ('Diana', 23, 88.0)
""")

# Commit the changes
conn.commit()

# Query data from the table
cursor.execute("SELECT * FROM students")
result = cursor.fetchall()
print("All students:")
print(result)

# Query with filtering
cursor.execute("SELECT name, grade FROM students WHERE grade > 85")
high_performers = cursor.fetchall()
print("\nHigh performers:")
print(high_performers)

# Query with aggregation
cursor.execute("SELECT AVG(grade) as average_grade FROM students")
avg_grade = cursor.fetchone()
print(f"\nAverage grade: {avg_grade[0]}")

# Close the connection when done
cursor.close()
conn.close()

# ============================================================================
# Example 2: Using pandas with SQLite
# ============================================================================

# Connect to the database
conn = sqlite3.connect('my_database.db')

# Read entire table into a DataFrame
students_df = pd.read_sql_query("SELECT * FROM students", conn)
print("\nStudents DataFrame:")
print(students_df)

# Query with filtering using pandas
high_performers_df = pd.read_sql_query(
    "SELECT name, grade FROM students WHERE grade > 85", 
    conn
)
print("\nHigh performers DataFrame:")
print(high_performers_df)

# Close connection
conn.close()

# ============================================================================
# Example 3: Using SQLAlchemy (Recommended for Complex Applications)
# ============================================================================

# Create an SQLAlchemy engine
engine = create_engine('sqlite:///my_database.db')

# Create a sample dataframe
students_data = pd.DataFrame({
    'id': [1, 2, 3, 4, 5],
    'name': ['Alice', 'Bob', 'Charlie', 'Diana', 'Eve'],
    'age': [20, 22, 21, 23, 20],
    'grade': [85.5, 92.0, 78.5, 88.0, 95.5]
})

# Write dataframe to database
students_data.to_sql('students', engine, if_exists='replace', index=False)

# Read entire table back
with engine.connect() as conn:
    students_from_db = pd.read_sql_table('students', conn)
    print("\nStudents from database:")
    print(students_from_db)

# Execute raw SQL with SQLAlchemy
with engine.connect() as conn:
    result = conn.execute(text("SELECT * FROM students WHERE age > 20"))
    print("\nStudents older than 20:")
    for row in result:
        print(row)

# Get list of tables
from sqlalchemy import inspect
inspector = inspect(engine)
tables = inspector.get_table_names()
print(f"\nTables in database: {tables}")

# ============================================================================
# Example 4: pandas DataFrame Operations with SQL
# ============================================================================

# Using pandas read_sql with more complex queries
with engine.connect() as conn:
    # Filtering and sorting
    query = """
        SELECT name, grade 
        FROM students 
        WHERE grade > 80 
        ORDER BY grade DESC
    """
    result_df = pd.read_sql_query(query, conn)
    print("\nFiltered and sorted:")
    print(result_df)

# ============================================================================
# Example 5: MySQL Database Connection
# ============================================================================

# For MySQL databases (requires MySQL server running)
# pip install pymysql

# import pymysql
# from sqlalchemy import create_engine
# 
# # Method 1: Using pymysql directly
# conn = pymysql.connect(
#     host='localhost',
#     port=3306,
#     user='username',
#     password='password',
#     database='my_database'
# )
# cursor = conn.cursor()
# cursor.execute("SELECT * FROM my_table")
# result = cursor.fetchall()
# conn.close()
# 
# # Method 2: Using SQLAlchemy
# engine = create_engine('mysql+pymysql://username:password@localhost:3306/my_database')
# df = pd.read_sql_query("SELECT * FROM my_table", engine)

# ============================================================================
# Example 6: PostgreSQL Database Connection
# ============================================================================

# For PostgreSQL databases (requires PostgreSQL server running)
# pip install psycopg2-binary

# import psycopg2
# from sqlalchemy import create_engine
# 
# # Method 1: Using psycopg2 directly
# conn = psycopg2.connect(
#     host='localhost',
#     port=5432,
#     user='username',
#     password='password',
#     database='my_database'
# )
# cursor = conn.cursor()
# cursor.execute("SELECT * FROM my_table")
# result = cursor.fetchall()
# conn.close()
# 
# # Method 2: Using SQLAlchemy
# engine = create_engine('postgresql://username:password@localhost:5432/my_database')
# df = pd.read_sql_query("SELECT * FROM my_table", engine)

# ============================================================================
# Common SQL Operations in Python
# ============================================================================

# Reconnect for examples
engine = create_engine('sqlite:///my_database.db')

# 1. SELECT - Retrieve data
with engine.connect() as conn:
    result = pd.read_sql_query("SELECT * FROM students", conn)
    print("\n1. SELECT all:")
    print(result)
    
    result = pd.read_sql_query("SELECT name, grade FROM students WHERE age > 20", conn)
    print("\n1. SELECT with WHERE:")
    print(result)

# 2. INSERT - Add new records
with engine.connect() as conn:
    conn.execute(text("INSERT INTO students (name, age, grade) VALUES ('Frank', 24, 89.0)"))
    conn.commit()
    print("\n2. INSERT: Added Frank")

# 3. UPDATE - Modify existing records
with engine.connect() as conn:
    conn.execute(text("UPDATE students SET grade = 90.0 WHERE name = 'Frank'"))
    conn.commit()
    print("\n3. UPDATE: Updated Frank's grade")

# 4. DELETE - Remove records
with engine.connect() as conn:
    conn.execute(text("DELETE FROM students WHERE name = 'Frank'"))
    conn.commit()
    print("\n4. DELETE: Removed Frank")

# 5. ORDER BY - Sort results
with engine.connect() as conn:
    result = pd.read_sql_query("SELECT * FROM students ORDER BY grade DESC", conn)
    print("\n5. ORDER BY grade DESC:")
    print(result)

# 6. GROUP BY - Aggregate data
with engine.connect() as conn:
    result = pd.read_sql_query(
        "SELECT age, AVG(grade) as avg_grade FROM students GROUP BY age", 
        conn
    )
    print("\n6. GROUP BY age:")
    print(result)

# 7. JOIN - Combine tables (example with two tables)
with engine.connect() as conn:
    # First create a courses table
    conn.execute(text("""
        CREATE TABLE IF NOT EXISTS courses (
            student_id INTEGER,
            course_name TEXT,
            credits INTEGER
        )
    """))
    
    conn.execute(text("""
        INSERT INTO courses (student_id, course_name, credits) VALUES 
        (1, 'Math', 3),
        (1, 'Physics', 4),
        (2, 'Chemistry', 3),
        (3, 'Biology', 4)
    """))
    conn.commit()
    
    # Join students and courses
    result = pd.read_sql_query("""
        SELECT s.name, c.course_name, c.credits 
        FROM students s 
        JOIN courses c ON s.id = c.student_id
    """, conn)
    print("\n7. JOIN students and courses:")
    print(result)

# 8. COUNT, SUM, AVG, MIN, MAX
with engine.connect() as conn:
    result = pd.read_sql_query("""
        SELECT 
            COUNT(*) as total_students,
            AVG(grade) as avg_grade,
            MIN(grade) as min_grade,
            MAX(grade) as max_grade
        FROM students
    """, conn)
    print("\n8. Aggregate functions:")
    print(result)

# ============================================================================
# Using Parameterized Queries (Prevent SQL Injection)
# ============================================================================

with engine.connect() as conn:
    # Parameterized query using SQLAlchemy
    name_to_find = 'Alice'
    result = pd.read_sql_query(
        text("SELECT * FROM students WHERE name = :name"),
        conn,
        params={'name': name_to_find}
    )
    print("\nParameterized query result:")
    print(result)

# ============================================================================
# Using Context Managers and Transactions
# ============================================================================

# Using context manager for automatic resource cleanup
with engine.connect() as conn:
    with conn.begin():  # Start transaction
        conn.execute(text("INSERT INTO students (name, age, grade) VALUES ('Grace', 22, 87.0)"))
        conn.execute(text("INSERT INTO students (name, age, grade) VALUES ('Henry', 21, 91.5)"))
        # Automatically commits if no exception, rolls back if exception occurs

print("\nTransaction completed")

# ============================================================================
# Using pandasql for SQL on DataFrames (Alternative to SQL databases)
# ============================================================================

# pip install pandasql

# from pandasql import sqldf
# 
# # Create sample dataframes
# students_df = pd.DataFrame({
#     'name': ['Alice', 'Bob', 'Charlie'],
#     'age': [20, 22, 21],
#     'grade': [85.5, 92.0, 78.5]
# })
# 
# # Run SQL query on the dataframe
# query = "SELECT name, grade FROM students_df WHERE grade > 80"
# result = sqldf(query, locals())
# print(result)

# ============================================================================
# Best Practices
# ============================================================================

# 1. Always close connections or use context managers
#    with engine.connect() as conn:
#        # operations
#    # Connection automatically closed

# 2. Use parameterized queries to prevent SQL injection
#    conn.execute(text("SELECT * FROM students WHERE name = :name"), {'name': name_var})

# 3. Use transactions for multiple related operations
#    with conn.begin():
#        # multiple operations
#        # automatically commits or rolls back

# 4. Use pandas for easy data manipulation
#    df = pd.read_sql_query(query, conn)

# 5. Use SQLAlchemy for database portability
#    Easy to switch between SQLite, MySQL, PostgreSQL, etc.

# 6. Handle exceptions properly
#    try:
#        # database operations
#    except Exception as e:
#        conn.rollback()
#        print(f"Error: {e}")

# ============================================================================
# Cleanup
# ============================================================================

# Close engine
engine.dispose()

# Optionally, remove the database file
# os.remove('my_database.db')

print("\n" + "="*80)
print("Summary")
print("="*80)
print("""
A database is basically a way of dealing efficiently with lots of 
potentially huge dataframes

SQL is the standard language for telling databases what to do, 
especially what queries to run

Everything in an SQL query is something we've practiced already in Python:
- Filtering (WHERE clause = boolean indexing)
- Aggregation (GROUP BY = groupby())
- Merging (JOIN = merge())
- Ordering (ORDER BY = sort_values())

Workflow:
1. Connect Python to the database
2. Send it an SQL query
3. Analyze the returned DataFrame

More information at: http://www.stat.cmu.edu/~cshalizi/statcomp/14/
""")
